packages/main/src/SplitButton.ts

```diff
@@ -275,66 +292,119 @@ class SplitButton extends UI5Element {
 		if (this.disabled) {
 			this._tabIndex = "-1";
 		}
 	}
 
+	onAfterRendering() {
+		if (this._isKeyDownOperation && !this.activeArrowButton) {
+			return;
+		}
+		this._setEffectiveActiveState();
+	}
+
+	_handleMouseClick(e: MouseEvent) {
+		const target = e.target as Button;
+
+		this._manageFocus(target);
+		this._fireClick(e);
+	}
+
 	_onFocusOut(e: FocusEvent) {
 		if (this.disabled || getEventMark(e)) {
 			return;
 		}
+
 		this._shiftOrEscapePressed = false;
-		this.focused = false;
 		this._setTabIndexValue();
+		this._manageFocus();
 	}
 
 	_onFocusIn(e: FocusEvent) {
 		if (this.disabled || getEventMark(e)) {
 			return;
 		}
 		this._shiftOrEscapePressed = false;
-		this.focused = true;
+		this._manageFocus(this);
+	}
+
+	_textButtonFocusIn(e?: FocusEvent) {
+		e?.stopPropagation();
+		this._manageFocus(this.textButton!);
+
+		this._setTabIndexValue();
 	}
 
 	_onKeyDown(e: KeyboardEvent) {
-		if (isDown(e) || isUp(e) || isDownAlt(e) || isUpAlt(e) || isF4(e)) {
-			e.preventDefault();
-			this._arrowButtonActive = true;
-			this._fireArrowClick();
-		} else if (isSpace(e) || isEnter(e)) {
-			e.preventDefault();
-			this._textButtonActive = true;
-			if (isEnter(e)) {
-				this._fireClick();
-			} else {
-				this._spacePressed = true;
-			}
+		this._isKeyDownOperation = true;
+		if (this._isArrowKeyAction(e)) {
+			this._handleArrowButtonAction(e);
+			this._activeArrowButton = true;
+		} else if (this._isDefaultAction(e)) {
+			this._handleDefaultAction(e);
+			this._isDefaultActionPressed = true;
+		}
+
+		if (this._spacePressed && this._isShiftOrEscape(e)) {
+			this._handleShiftOrEscapePressed();
 		}
-		if (this._spacePressed && (isEscape(e) || isShift(e))) {
-			this._shiftOrEscapePressed = true;
+
+		// Handles button freeze issue when pressing Enter/Space and navigating with Tab/Shift+Tab simultaneously.
+		if (this._isDefaultActionPressed && (isTabNext(e) || isTabPrevious(e))) {
+			this._activeArrowButton = false;
 			this._textButtonActive = false;
 		}
 
-		this._setTabIndexValue();
+		this._tabIndex = "-1";
 	}
 
 	_onKeyUp(e: KeyboardEvent) {
-		if (isDown(e) || isUp(e) || isDownAlt(e) || isUpAlt(e) || isF4(e)) {
-			this._arrowButtonActive = false;
-		} else if (isSpace(e) || isEnter(e)) {
+		this._isKeyDownOperation = false;
+		if (this._isArrowKeyAction(e)) {
+			e.preventDefault();
+			this._activeArrowButton = false;
 			this._textButtonActive = false;
+		} else if (this._isDefaultAction(e)) {
+			this._isDefaultActionPressed = false;
+			this._textButtonActive = false;
+			this._activeArrowButton = false;
 			if (isSpace(e)) {
 				e.preventDefault();
 				e.stopPropagation();
 				this._fireClick();
 				this._spacePressed = false;
+				this._textButtonActive = false;
+				this.arrowButton!._keepActiveState = false;
 			}
 		}
 
-		this._setTabIndexValue();
+		if (this._isShiftOrEscape(e)) {
+			if (this.activeArrowButton) {
+				this._activeArrowButton = true;
+				this.arrowButton!._keepActiveState = true;
```

Comment:
After our discussion we talked that we could either use `_setEffectiveActiveState` or it might turn out that those lines are redundant as we'll hit the same function in the onAfterRendering hook. We have the same situation in the other handlers.